language/dsl/src/types.ts

```diff
@@ -56,12 +57,21 @@ export type SwapKeysToType<T, K extends keyof T, NewType> = {
   [P in keyof T]: P extends K ? NewType : T[P];
 };
 
+/**
+ * Note: have to explicitly handle boolean cases, otherwise boolean gets distributed to
+ * the union as true | false breaking the mapping.
+ */
 export type WithTemplateTypes<T> =
   T extends Record<any, any>
     ? {
         [P in keyof T]: WithTemplateTypes<T[P]>;
       }
-    : T | BindingTemplateInstance | ExpressionTemplateInstance;
+    : T extends boolean
+      ?
+          | boolean
+          | BindingTemplateInstance<boolean>
+          | ExpressionTemplateInstance<boolean>
+      : T | BindingTemplateInstance<T> | ExpressionTemplateInstance<T>;
```

Comment:
Could we decompose this a little bit to make it easier to understand? Maybe just one type for:
```ts
          | boolean
          | BindingTemplateInstance<boolean>
          | ExpressionTemplateInstance<boolean>
```